_posts/2016-10-06-BDD-tests-for-Hadoop-with-Cucumber-part-I.md

```diff
@@ -0,0 +1,310 @@
+---
+title: BDD tests for Hadoop with Cucumber. Part I
+description: "Tutorial how to create behaviour tests for Hadoop application with Cucumber. Part I: introduction and environment"
+layout: post
+author: bogdan
+tags: [bog data, java, hadoop, qa]
+modified: 2016-10-06
+image:
+    feature: posts/bdd_test_hadoop/title.jpg
+    
+---
+
+Big Data is a process of a huge amount of structured (more or less) data that should be filtered or sorted for future analysis or, 
+actually made analysis. So much data that you can’t process it on the one computer for a reasonable time. 
+This can be logs from services, like call service, or logs from web servers that contain billions of records, 
+so you should process data in parallel  on a bunch of computers. Do do it so we use specific software based on MapReduce pattern. 
+In our case, we use Hadoop with Cascading framework. Hadoop implement MapReduce and Cascading framework contain a lot of useful tools, 
+for example abstraction to Amazon S3, so we can use S3 buckets as input/output folders for processing jobs, 
+but in the same time we allow to use regular file systems on developer machines in development and test purposes.
+
+As any other piece of software it should be tested before using in production, especially because logical mistakes in processing a lot of data cost a much. 
+Unfortunately there are no much information about BDD testing the Hadoop and Hive jobs. 
+So i decide to write how we do it in our Cascading project here, in Intelliarts.
+
+<!-- more -->
+
+# What do we have
+
+Assume we have a lot of data about phone calls in bunch of files. File contain lines with following format: `Id,country,time,duration` 
+Where `id` is call unique id, `country` is a caller country, `time` is a time when call start (timestamp) and `duration` is a call duration in milliseconds. 
+
+For example: 
+```
+0000,UA,1433998201,60000
+0001,US,1433998201,30000
+0001,GB,1433998301,30000
+...
+```
+
+# What we want to do?
+
+Let's sort calls from these records by country (each country will be in separate folder) and change the format of records - replace delimiters from `,` to tab symbol `\t`. 
+From the date of previous example so we should get the folders `US`, `GB`, `UA` and `UA` will contain file with records looks like this:
+
+```
+id	country	time	duration
+0000	UA	1433998201	60000
+```
+
+And, we want to be sure that we did any mistakes during implementation, so we want to test the results. 
```

Comment:
And, we want to be sure that we did any mistakes during implementation
And, we want to be sure that we did **not make** any mistakes during implementation